SingleSource/UnitTests/Atomic/float_test.cpp

```diff
@@ -0,0 +1,135 @@
+//===--- float_test.cc -- Testing aligned floating point numbers -- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// The following text is present in each test file:
+//
+// These tests aim to capture real-world multithreaded use cases of atomic
+// builtins. Each test focuses on a single atomic operation. Those using
+// multiple operations can be compared with other tests using the same
+// operations to isolate bugs to a single atomic operation.
+//
+// Each test consists of a "looper" body and a test script. The test script
+// instantiates 10 threads, each running the looper. The loopers contend the
+// same memory address, performing atomic operations on it. Each looper executes
+// 10^6 times for a total of 10^7 operations. The resultant value in the
+// contended pointer is compared against a closed-form solution. It's expected
+// that the two values equate.
+//
+// For example, a looper that increments the shared pointer is expected to end
+// up with a value of 10^7. If its final value is not that, the test fails.
+//
+// Each test also tests the corresponding nonatomic operation with a separate
+// shared variable. Ideally, this value differs from the atomic "correct" value,
+// and the test can compare the two. In reality, some simpler operations (e.g.
+// those conducted through the ALU) can still end up with the correct answer,
+// even when performed nonatomically. These tests do not check the nonatomic
+// result, although it is still outputted to aid in debugging.
+//
+// Each test is performed on all relevant types.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file tests atomic operations on floating point types with aligned
+// memory addresses.
+//
+// The types tested are: float, double.
+// The ops tested are: xchg, cmpxchg.
+//
+//===----------------------------------------------------------------------===//
+
+#include <sys/stat.h>
+
+#include <cstdio>
+#include <thread>
+#include <vector>
+
+#include "numeric.h"
+#include "util.h"
+
+// See numeric.h for an explanation of numeric xchg tests.
+template <typename T>
+void test_float_scalar_xchg() {
+  static constexpr T val = V >> right_shift<T>();
+  static constexpr T expected = val * kExpected;
+  std::vector<std::thread> pool;
+
+  for (int model : atomic_exchange_models) {
+    T afloat = 0;
+    T ffloat = 0;
+    for (int n = 0; n < kThreads; ++n) {
+      pool.emplace_back(looper_numeric_xchg_atomic<T>, &afloat, model);
```

Comment:
Nit: llvm style https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements prefer to drop the braces here.

IIRC Google style clang-format often places the body on the same line, thus many people prefer `{}` to work around it.